File: backend/models/post.py
# backend/models/post.py
"""Post data models."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class Post(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    author_uid: str
    author_username: str = "unknown"
    author_display_name: str = "Anonymous"
    author_photo_url: str = ""
    text: str
    image_url: str | None = None
    media_urls: list[str] = Field(default_factory=list)
    media_type: str = "text"
    status: Literal["approved", "rejected", "pending_review"] = "approved"
    moderation_layer: str | None = None
    moderation_reason: str | None = None
    rejection_reason: str | None = None  # human-readable reason shown to author
    flagged_terms: list[str] = Field(default_factory=list)
    like_count: int = 0
    comment_count: int = 0
    view_count: int = 0
    created_at: datetime
    updated_at: datetime
    schema_version: int = 1


class CreatePostRequest(BaseModel):
    text: str = Field(..., min_length=1, max_length=2000)
    media_urls: list[str] = Field(default_factory=list, max_length=10)
    media_type: str = "text"
    # When True, knowingly submit flagged content for human verification
    # (stored as pending_review) instead of being rejected with a 422.
    submit_for_review: bool = False

File: frontend/lib/shared/models/post.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'post.freezed.dart';
part 'post.g.dart';

@freezed
class Post with _$Post {
  const factory Post({
    required String id,
    @JsonKey(name: 'author_uid') required String authorUid,
    @JsonKey(name: 'author_username') required String authorUsername,
    @JsonKey(name: 'author_display_name') required String authorDisplayName,
    @JsonKey(name: 'author_photo_url') required String authorPhotoUrl,
    required String caption,
    @JsonKey(name: 'media_urls') @Default([]) List<String> mediaUrls,
    @JsonKey(name: 'media_type') required String mediaType,
    required String status,
    @JsonKey(name: 'like_count') @Default(0) int likeCount,
    @JsonKey(name: 'comment_count') @Default(0) int commentCount,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    @JsonKey(name: 'is_liked') @Default(false) bool isLiked,
  }) = _Post;

  factory Post.fromJson(Map<String, dynamic> json) => _$PostFromJson(json);
}
